source/libuv/uv_event_loop.c

```diff
@@ -0,0 +1,819 @@
+    while (open_subs_it != open_subs_end) {
+        struct handle_data *handle_data = AWS_CONTAINER_OF(open_subs_it, struct handle_data, open_subs_node);
+        if (uv_poll_start(&handle_data->poll, handle_data->uv_events, s_uv_poll_cb)) {
+            goto clean_up;
+        }
+        aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+        open_subs_it = open_subs_it->next;
+    }
+    cleanup_polls = true;
+
+    /* Start all existing timers */
+    aws_hash_table_foreach(&impl->on_thread_data.running_tasks, s_running_tasks_start, impl);
+
+    if (impl->owns_uv_loop) {
+        assert(s_owned(impl)->state == EVENT_THREAD_STATE_READY_TO_RUN);
+
+        if (aws_thread_launch(&s_owned(impl)->thread, &s_thread_loop, impl, NULL)) {
+            goto clean_up;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+
+clean_up:
+
+    if (cleanup_polls) {
+        struct aws_linked_list_node *open_subs_it = aws_linked_list_begin(&impl->on_thread_data.open_subscriptions);
+        const struct aws_linked_list_node *open_subs_end = aws_linked_list_end(&impl->on_thread_data.open_subscriptions);
+        while (open_subs_it != open_subs_end) {
+            struct handle_data *handle_data = AWS_CONTAINER_OF(open_subs_it, struct handle_data, open_subs_node);
+
+            uv_close((uv_handle_t *)&handle_data->poll, s_uv_close_sub);
+
+            open_subs_it = open_subs_it->next;
+        }
+    }
+    if (cleanup_async_stop) {
+        uv_close((uv_handle_t *)&impl->stop_async, s_uv_close_handle);
+    }
+    if (cleanup_async_schedule_tasks) {
+        uv_close((uv_handle_t *)&impl->cross_thread_data, s_uv_close_handle);
+    }
+
+    return AWS_OP_ERR;
+}
+
+static int s_stop(struct aws_event_loop *event_loop) {
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    if (uv_async_send(&impl->stop_async)) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_wait_for_stop_completion(struct aws_event_loop *event_loop) {
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    int status = AWS_OP_SUCCESS;
+
+    if (impl->owns_uv_loop) {
+        status = aws_thread_join(&s_owned(impl)->thread);
+    }
+
+    /* Wait for all handles to close */
+    while (aws_atomic_load_int(&impl->num_open_handles)) { }
+
+    return status;
+}
+
+static void s_schedule_task_common(struct aws_event_loop *event_loop, struct aws_task *task, uint64_t run_at_nanos) {
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    task->timestamp = run_at_nanos;
+
+    /* Begin critical section */
+    aws_mutex_lock(&impl->cross_thread_data.mutex);
+
+    bool should_signal_thread = aws_linked_list_empty(&impl->cross_thread_data.tasks_to_schedule);
+    aws_linked_list_push_back(&impl->cross_thread_data.tasks_to_schedule, &task->node);
+
+    /* Signal thread that cross_thread_data has changed (unless it's been signaled already) */
+    if (should_signal_thread) {
+        uv_async_send(&impl->schedule_tasks_async);
```

Comment:
Does this need to be inside the mutex? Is it also protecting/serializing uv_async_send()?

Reply:
No, async send ops are thread safe.